[Review] Add tests for dom-class

steveoh opened the pull request:
Can I get some feedback on my direction so far. There are no DOH tests that I can find so I'm making a go of this from looking at other tests. Is this a worth while effort or a duplicate of effort? 

Comments appreciated.

wkeese commented:
The DOH tests are at https://github.com/dojo/dojo/blob/1.10/tests/_base/html.html#L246.  Looks like they got removed from the tree prematurely (i.e. before they were replaced by Intern tests).

I didn't compare your new tests to the old DOH ones, but at a glance the new tests seem good.   I'll add some line item comments if I see issues.

vansimke commented:
@wkeese, I think those tests are still present. Check this link out from master: https://github.com/dojo/dojo/blob/master/testsDOH/_base/html.html#L223

I didn't get a chance to convert all of the tests during my blitz a few months back, but I tried to be very careful about not removing tests that weren't moved over.

@steveoh, a lot of the test in things like dom-class were already done in the _base/html.html tests. I think that there must have been a major refactor (probably ~1.7) that triggered many global functions to be pulled out into AMD modules. When this happened, the global functions were just turned into aliases for the modules. As a result, they are tested, but in a bit of a round about manner. I tried to update the test to verify that the API of the module was working as expected (often with new tests), then, on Bill's advice, I pulled in the legacy tests from _base/html.html into a section of "validation tests" so that no edge tests were lost. Check out this: https://github.com/dojo/dojo/blob/master/tests/unit/dom-style.js for an example of how I approached this.

steveoh commented:
so @wkeese and @vansimke it sounds like this effort is still worth while. I can include the verification tests for the old doh tests for posterity, and continue with my direction. Does that seems ok for everyone or should I switch gears?

wkeese commented:
Sounds worthwhile to me.  Probably you don't need to include old tests "for posterity" but you should just read through the old tests to make sure that you didn't miss anything.

steveoh commented:
@wkeese do you want me to squash or make any other changes before calling this final?

wkeese commented:
Personally it seems wrong to test undefined behavior and expect a certain result, such as calling `domClass.add(node)` and checking that "nothing happens".   I don't see the point of that kind of test, and conversely (theoretically), maybe in the future that call will throw an exception, which would cause your test to spuriously fail.

steveoh commented:
@wkeese i removed those tests. :shipit:?

wkeese commented on tests/unit/dom-class.js:
```diff
@@ -0,0 +1,840 @@
+            },
+
+            "node + class + false": function () {
+                // when you pass a node and class and a false condition, the class is added
```
you meant "removed" not "added".   likewise for below.

wkeese commented on tests/unit/dom-class.js:
```diff
@@ -0,0 +1,840 @@
+                             domClass.contains(node, multipleCss);
+
+                assert.isTrue(result);
+
```
Not important, but did you know that you can have multiple assert() statements in your function?   So I think most people would write this like:

```
assert.isTrue(domClass.contains(node, cssClass));
assert.isTrue(domClass.contains(node, multipleCss));
```

(I also don't know why you keep setting a result variable that you only use once.)

wkeese commented:
> @wkeese i removed those tests. Merge it?

Well, it still looks like there are a bunch to me.   For example:

```
domClass.remove(nodeId, undefined);
domClass.replace(node);
domClass.toggle(node);
```

steveoh commented:
@wkeese from [the docs](http://dojotoolkit.org/reference-guide/1.10/dojo/dom-class.html)

````js
domClass.remove(nodeId, undefined);
````
> If the second argument is not specified or undefined, all classes will be removed.

I removed the other undocumented use case tests. I split out the asserts where it made sense and updated the comment text. I also removed the unnecessary `results` variable. 

Let me know if you have any other issues with this.

To run these tests 
1. cd into dojo
1. npm run test-proxy
1. browse to http://localhost:9001/__intern/client.html?config=tests/dojo.intern#

wkeese commented:
OK, thanks, I didn't realize the behavior of `domClass.remove(nodeId, undefined);` documented.

I just looked over  https://github.com/dojo/dojo/blob/1.10/tests/_base/html.html#L246.   It seems like it's testing much more complex cases than you've covered.  For example, this code block:

```js
dojo.addClass(node, "a");
t.is("a", node.className, "class is a");
dojo.addClass(node, "a b");
t.is("a b", node.className, "class is a b");
dojo.addClass(node, "b a");
t.is("a b", node.className, "class is still a b");
```

It's checking:

1. specifying multiple classes as a space separated string (including testing specifying the classes in a different order (i.e. "b a"), to make sure that "b a" is treated as two separate classes rather than one class name with an embedded space)
2. adding classes that already exist to make sure that doesn't produce duplicates

The DOM tests have similar checks for other dom-class methods.  Did you cover that kind of thing?   I didn't notice it, but maybe it's there.

steveoh commented:
ok @wkeese. I added all of those tests for posterity into the regression suite. That way we won't miss anything. Let me know what else is necessary.

steveoh commented:
@wkeese I fixed the mixed up `assert.equals` and [renamed the tests to match the others](https://github.com/dojo/dojo/blob/master/tests/unit/dom-style.js#L411). 

Once again, let me know what else is necessary.

wkeese commented:
Thanks, seems reasonable.  I was about to push this, but I don't see a CLA (http://dojofoundation.org/about/cla) for you?

steveoh commented:
I scanned one of the paper ones in a long time ago. I just filled out the
new form.

wkeese commented:
OK thanks!   I pushed your test as f29befa2a44394c3221f403fda19ed16c199e114.   It's what you submitted here, except I had to remove some trailing comma syntax errors so that the test will run on IE8.  I also removed a bunch of blank lines at the end of functions.
